src/Xamarin.Android.Build.Tasks/Utilities/ApplicationConfigNativeAssemblyGenerator.cs

```diff
@@ -16,8 +21,77 @@ enum MonoComponent
 		Tracing   = 0x04,
 	}
 
-	class ApplicationConfigNativeAssemblyGenerator : NativeAssemblyGenerator
+	class ApplicationConfigNativeAssemblyGenerator : NativeAssemblyComposer
 	{
+		sealed class DSOCacheEntryContextDataProvider : NativeAssemblerStructContextDataProvider
+		{
+			public override string GetComment (object data, string fieldName)
+			{
+				var dso_entry = data as DSOCacheEntry;
+				if (dso_entry == null) {
+					throw new InvalidOperationException ("Invalid data type, expected an instance of DSOCacheEntry");
+				}
+
+				if (String.Compare ("hash", fieldName, StringComparison.Ordinal) == 0) {
+					return $"hash, from name: {dso_entry.HashedName}";
+				}
+
+				if (String.Compare ("name", fieldName, StringComparison.Ordinal) == 0) {
+					return $"name: {dso_entry.name}";
+				}
+
+				return String.Empty;
+			}
+		}
+
+		[NativeAssemblerStructContextDataProvider (typeof (DSOCacheEntryContextDataProvider))]
+		sealed class DSOCacheEntry
+		{
+			[NativeAssembler (Ignore = true)]
+			public string HashedName;
+
+			[NativeAssembler (UsesDataProvider = true)]
+			public ulong hash;
+			public bool ignore;
+
+			[NativeAssemblerString (UsesDataProvider = true)]
+			public string name;
+			public IntPtr handle = IntPtr.Zero;
+		}
+
+		// Order of fields and their type must correspond *exactly* to that in
+		// src/monodroid/jni/xamarin-app.hh AssemblyStoreSingleAssemblyRuntimeData structure
+		sealed class AssemblyStoreSingleAssemblyRuntimeData
```

Comment:
What is the *relationship* here?  Do these types need to be kept in-sync *manually*?  Is a tool involved?  Something else?

Reply:
Nothing has changed in this regard yet. Still maintained manually.